radix-engine/src/track/db_key_mapper.rs

```diff
@@ -0,0 +1,88 @@
+use radix_engine_common::data::scrypto::{scrypto_decode, ScryptoDecode};
+use radix_engine_interface::crypto::{hash, Hash};
+use radix_engine_interface::types::{ModuleId, NodeId, SubstateKey};
+use radix_engine_store_interface::interface::{
+    DbPartitionKey, DbSortKey, PartitionEntry, SubstateDatabase,
+};
+use sbor::rust::prelude::*;
+
+/// A mapper between the business RE Node / Module / Substate IDs and database keys.
+pub trait DatabaseKeyMapper {
+    /// Converts the given RE Node and Module ID to the database partition's key.
+    fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey;
+
+    /// Converts the given Substate key to the database's sort key.
+    fn to_db_sort_key(key: &SubstateKey) -> DbSortKey;
+}
+
+/// A [`DatabaseKeyMapper`] tailored for databases which cannot tolerate long common prefixes
+/// among keys (for performance reasons). In other words, it spreads the keys "evenly" (i.e.
+/// pseudo-randomly) across the key space.
+/// For context: our actual use-case for it is the Jellyfish Merkle Tree.
+pub struct SpreadPrefixKeyMapper;
+
+impl DatabaseKeyMapper for SpreadPrefixKeyMapper {
+    fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey {
+        let mut buffer = Vec::new();
+        buffer.extend(node_id.as_ref());
+        buffer.push(module_id.0);
+        let hash_bytes = hash(buffer).0[(Hash::LENGTH - 26)..Hash::LENGTH].to_vec(); // 26 bytes
+        DbPartitionKey(hash_bytes)
+    }
+
+    fn to_db_sort_key(key: &SubstateKey) -> DbSortKey {
+        let bytes = match key {
+            SubstateKey::Tuple(field) => {
+                vec![*field]
+            }
+            SubstateKey::Map(key) => {
+                hash(key).0[12..Hash::LENGTH].to_vec() // 20 bytes
+            }
+            SubstateKey::Sorted((bucket, key)) => {
+                let mut bytes = bucket.to_be_bytes().to_vec();
+                bytes.extend(hash(key).0[12..Hash::LENGTH].to_vec()); // 20 bytes
+                bytes
+            }
+        };
+        DbSortKey(bytes)
+    }
+}
+
+/// Convenience methods for direct `SubstateDatabase` readers.
+pub trait MappedSubstateDatabase {
```

Comment:
I still feel weird about us needing to specify `SpreadPrefixKeyMapper` in every function call - it keeps making me feel like the mapper choice is arbitrary when it's really not.

I wonder if we specify it as a generic parameter to the trait and remove it from the methods?
EG: `MappedSubstateDatabase<M: DatabaseKeyMapper>` and then only implement specifically `impl<S: SubstateDatabase> MappedSubstateDatabase<SpreadPrefixKeyMapper> for S` then I imagine we can drop the specification of M everywhere - what do you think? Rust might complain about it though, but I don't think it would.

My suggestion feels a bit weird though.

Alternatively:
* We can just fix `pub use SpreadPrefixKeyMapper as ProtocolKeyMapper`, use that in the `MappedSubstateDatabase` impl and let it be changed at compile time if people wanted to 🤷 .
* We can have a `get_protocol_mapped_substate` method on a new trait which fixes `M = ProtocolKeyMapper = SpreadPrefixKeyMapper` and just use that method everywhere :')

Reply:
hah, look, [the 8th time](https://github.com/radixdlt/radixdlt-scrypto/pull/986#discussion_r1181195620) :p

let's drop the trait?